R/verb-pivot-longer.R

```diff
@@ -0,0 +1,302 @@
+#' @param names_to A string specifying the name of the column to create
+#'   from the data stored in the column names of `data`.
+#' @param names_prefix A regular expression used to remove matching text
+#'   from the start of each variable name.
+#' @param names_sep,names_pattern If `names_to` contains multiple values,
+#'   these arguments control how the column name is broken up.
+#' @param names_repair What happens if the output has invalid column names?
+#' @param values_to A string specifying the name of the column to create
+#'   from the data stored in cell values. If `names_to` is a character
+#'   containing the special `.value` sentinel, this value will be ignored,
+#'   and the name of the value column will be derived from part of the
+#'   existing column names.
+#' @param values_drop_na If `TRUE`, will drop rows that contain only `NA`s
+#'   in the `value_to` column.
+#' @param names_transform,values_transform A list of column name-function pairs.
+#' @param names_ptypes A list of column name-prototype pairs.
+#' @param values_ptypes Not supported.
+#' @param ... Additional arguments passed on to methods.
+#' @examples
+#' # See vignette("pivot") for examples and explanation
+#'
+#' # Simplest case where column names are character data
+#' if (require("tidyr", quietly = TRUE)) {
+#'   memdb_frame(
+#'     id = c("a", "b"),
+#'     x = 1:2,
+#'     y = 3:4
+#'   ) %>%
+#'     pivot_longer(-id)
+#' }
+pivot_longer.tbl_lazy <- function(data,
+                                  cols,
+                                  names_to = "name",
+                                  names_prefix = NULL,
+                                  names_sep = NULL,
+                                  names_pattern = NULL,
+                                  names_ptypes = list(),
+                                  names_transform = list(),
+                                  names_repair = "check_unique",
+                                  values_to = "value",
+                                  values_drop_na = FALSE,
+                                  values_ptypes,
+                                  values_transform = list(),
+                                  ...) {
+  if (!is_missing(values_ptypes)) {
+    abort("The `values_ptypes` argument is not supported for remote back-ends")
+  }
+
+  ellipsis::check_dots_empty()
+
+  cols <- enquo(cols)
+  spec <- dbplyr_build_longer_spec(data, !!cols,
+    names_to = names_to,
+    values_to = values_to,
+    names_prefix = names_prefix,
+    names_sep = names_sep,
+    names_pattern = names_pattern,
+    names_ptypes = names_ptypes,
+    names_transform = names_transform
+  )
+
+  dbplyr_pivot_longer_spec(data, spec,
+    names_repair = names_repair,
+    values_drop_na = values_drop_na,
+    values_transform = values_transform
+  )
+}
+
+# this is a nearly identical copy of `tidyr::build_longer_spec` adapted to
+# work with `tbl_lazy` objects.
+dbplyr_build_longer_spec <- function(data, cols,
```

Comment:
Do you think there's some way that we could use `tidyr::build_longer_spec()`, possibly by creating a dummy dataframe that matches the structure of the database table? I've admittedly only skimmed `dbplyr_build_longer_spec()`, but I don't see anything obviously database specific.

Reply:
Yes, it actually was trivial.